tests/evaluator_tests.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+import numpy
+import unittest
+from lib.collaborative_filtering import CollaborativeFiltering
+from lib.evaluator import Evaluator
+from util.data_parser import DataParser
+from util.model_initializer import ModelInitializer
+
+
+class TestcaseBase(unittest.TestCase):
+    def setUp(self):
+        """
+        Setup method that is called at the beginning of each test.
+        """
+        self.documents, self.users = 8, 10
+        documents_cnt, users_cnt = self.documents, self.users
+        self.config = {'n_factors': 5, '_lambda': 0.01}
+        self.n_iterations = 15
+        self.initializer = ModelInitializer(self.config.copy(), self.n_iterations)
+
+        self.n_recommendations = 1
+
+        def mock_get_ratings_matrix(self=None):
+
+            return [[int(not bool((article + user) % 3)) for article in range(documents_cnt)]
+                    for user in range(users_cnt)]
+        self.ratings_matrix = numpy.array(mock_get_ratings_matrix())
+        setattr(DataParser, "get_ratings_matrix", mock_get_ratings_matrix)
+
+        evaluator = Evaluator(self.ratings_matrix)
+        collaborative_filtering = CollaborativeFiltering(self.initializer, self.n_iterations,
+                                                         self.ratings_matrix, evaluator,
+                                                         self.config, load_matrices=True)
+        collaborative_filtering.train()
+        self.predictions = (collaborative_filtering.get_predictions())
+        self.rounded_predictions = (collaborative_filtering.rounded_predictions())
+
+
+class TestEvaluator(TestcaseBase):
+    def runTest(self):
+        evaluator = Evaluator(self.ratings_matrix)
+        self.assertEqual(self.predictions.shape, self.ratings_matrix.shape)
+        recall_at_x = evaluator.recall_at_x(self.n_recommendations, self.predictions)
+        # if predictions are  perfect
+        if recall_at_x == 1:
+            for row in range(self.users):
+                for col in range(self.documents):
+                    self.assertEqual(self.rounded_predictions[row, col], self.ratings_matrix[row, col])
+
+        # If we modify all the top predictions for half the users,
+        # recall should be 0.5 by definition
+        for i in range(0, self.users, 2):
+            evaluator.ratings[i, (numpy.argmax(self.predictions[i], axis=0))] = 0
+        recall_at_x = evaluator.recall_at_x(self.n_recommendations, self.predictions)
+        self.assertEqual(0.5, recall_at_x)
+
+        self.setUp()
```

Comment:
The setup function is automatically called before any other function is ran, why is this necessary?